bctl/agent/agentservice_windows.go

```diff
@@ -0,0 +1,134 @@
+	agent.logger.Debugf("System service is %s", service.Platform())
+
+	options := make(service.KeyValue)
+	options["StartType"] = "automatic"
+	options["OnFailure"] = "restart"
+	svcConfig := &service.Config{
+		Name:        "BZAgent",
+		DisplayName: "BastionZero Agent Service",
+		Description: "This is a service responsible for the lifecycle of the BastionZero Agent.",
+		Option:      options,
+	}
+
+	if agentService.kardianosService, err = service.New(agentService, svcConfig); err != nil {
+		agent.logger.Errorf("failed to start agent service: %s", err)
+		return nil
+	}
+
+	errs := make(chan error, 5)
+	if agentService.serviceLogger, err = agentService.kardianosService.Logger(errs); err != nil {
+		agent.logger.Errorf("failed to start agent service logger: %s", err)
+		return nil
+	}
+
+	go func() {
+		for {
+			err := <-errs
+			if err != nil {
+				agent.logger.Errorf("service runtime error: %s", err)
+			}
+		}
+	}()
+
+	if len(svcFlag) != 0 {
+		err = service.Control(agentService.kardianosService, svcFlag)
+		if err != nil {
+			agent.logger.Error(err)
+			return nil
+		}
+		return agentService
+	}
+
+	if setupService {
+		err = service.Control(agentService.kardianosService, "install")
+		if err != nil {
+			agent.logger.Error(err)
+			return nil
+		}
+
+		err = service.Control(agentService.kardianosService, "start")
+		if err != nil {
+			agent.logger.Error(err)
+			return nil
+		}
+	}
+
+	return agentService
+}
+
+// Starts the agent service, required to implement the [service interface]: https://github.com/kardianos/service/blob/9832e01049dd49bb66dd8fc447c72eefba7fc2cd/service.go#L338
+func (as *AgentService) Start(s service.Service) error {
+	as.agent.logger.Info("Agent Service is Starting")
+	if service.Interactive() {
+		as.agent.logger.Info("Running in terminal.")
+	} else {
+		as.agent.logger.Info("Running under service manager.")
+	}
+
+	// Start should not block. Do the actual work async.
+	go as.agent.Run()
+	as.agent.logger.Info("Agent Service Started")
+	return nil
+}
+
+// Stops the agent service, required to implement the [service interface]: https://github.com/kardianos/service/blob/9832e01049dd49bb66dd8fc447c72eefba7fc2cd/service.go#L338
+func (as *AgentService) Stop(s service.Service) error {
+	as.agent.logger.Infof("Agent Service is Stopping")
+	as.agent.Close(&ServiceStopError{})
+	return nil
+}
+
+// Stops the agent service, required to implement the [service interface]: https://github.com/kardianos/service/blob/9832e01049dd49bb66dd8fc447c72eefba7fc2cd/service.go#L338
+func (as *AgentService) Run() error {
+
+	errChan := make(chan error)
+
+	// Listen to agent restarts and exit so the service manager can pick up from here and restart
+	go func() {
+		<-as.agent.Done()
+		var serviceStopErr *ServiceStopError
+		// If this is the service manager stopping the agent, just return and let the manager exit properly
+		if errors.As(as.agent.Err(), &serviceStopErr) {
+			errChan <- nil
```

Comment:
@lgmugnier I think we nearly had it, but this should `return` from its goroutine instead of sending to the `eerChan`. The problem is that the agent's tomb gets killed before we've closed the `controlConn`, so this channel send causes `main` to exit before `as.kardianosService.Run()` is done. I don't fully understand why this causes the agent to restart when it shouldn't, but I can see in the logs that it doesn't closed properly when we exit this function early like this.

I think the easiest move is just to exit this goroutine, let `Run()` finish, and return 0, stopping the ~Avatar~ restart cycle.

```suggestion
			return
```

Reply:
Adjusted here with the rest of changes
[52de01b](https://github.com/bastionzero/bzero/pull/233/commits/52de01b04b87720ce64e58b804f63fdb95a63587)